# First Script

```r
##########################
library(plyr)
#Read the table from magic-blast in
tab=read.table("betaglu_females",header=F)
#fixing column names
col_names=c("query.acc.","reference.acc.","% identity","not used","not.used",
            "not.used","query.start","query.end","reference.start","reference.end","not.used",
            "not.used","score","query.strand","reference.strand","query.length","BTOP",
            "num.placements","not.used","compartment","left.overhang","right.overhang",
            "mate.reference","mate.ref..start","composite.score")
colnames(tab)=col_names
#remove duplicate hits for eahc sequence
tab$seq=sub("..$","",tab$query.acc.)
tab2=tab[!duplicated(tab[,"seq"]),]
#clean SRA names to aggregate them
tab2$SRA=sub("\\..*","",tab2$seq)
#tab2$length=abs(tab2$query.start-tab2$query.end)
#tab2=tab2[tab2$length>=50,]
#tab2=tab2[tab2$X..identity>=90,]
#count freq of each enzyme for SRA
df <- count(tab2, c('SRA','reference.acc.'))
#save the clean table
# write.csv(df,"tablegene.csv")
#fuction for retrieving SRA MBS size
library(rentrez)
getSize <- function(ids) {size_mega <- c()
  for(i in ids) {term =  paste0(i, "[ACCN]")
    run = entrez_search(db = "sra", term = term)
    exp_descrip = entrez_summary(db = "sra", id = run[[1]])
    x = exp_descrip$run
    ## extract range defined by "total bases" and "load_done"
    size = substr(x, start = regexpr("total_bases=", x)[[1]][1] + attr(regexpr("total_bases=", x), "match.length"),stop = regexpr(" load_done", x)[[1]][1])
    ### some clean up
    size = gsub('\"', "", size, fixed = TRUE)
    ## converto to numeric and Mb
    size = as.numeric(size)/1e6
    size_mega <- c(size_mega, size)}
size_mega}
#retrieve the size of SRAs
df$MBS=getSize(df$SRA)
#Noprmalize using the size of each SRA
df$count=(df$freq*mean(df$MBS)/df$MBS)
#Transform to log10
df$trans_count=log2(df$count)
#Divide in tertiles or percentiles
df$group <- as.numeric(cut(df$trans_count, 4))

#Plotting boxplot of the pecentiles
g <- ggplot(subset(df, group %in% c(1, 2, 3)), aes(group, trans_count))
g + geom_violin(aes(fill=factor(group)))+ scale_fill_brewer(palette="Set1")+guides(fill=FALSE)+
   labs(x="Female Gut Metagenome",y="log2 of Beta-Glucuronidase Hits")+theme_classic(base_size = 30)+
   geom_point(color="black", position = "jitter")+scale_x_continuous(breaks=c(1,2,3),labels=c("Low","Medium","High"))

#Plotting boxplot of the pecentiles
g <- ggplot(df, aes(group, trans_count))
g + geom_violin(aes(fill=factor(group)))+ scale_fill_brewer(palette="Set1")+guides(fill=FALSE)+
   labs(x="Female Gut Metagenome",y="log2 of Beta-Glucuronidase Hits")+theme_classic(base_size = 30)+
   geom_point()+scale_x_continuous(breaks=c(1,2,3,4),labels=c("Low","Medium","High","Very High"))

#Plotting boxplot of the pecentiles
df$group <- as.numeric(cut(df$trans_count, 3))
g <- ggplot(df, aes(group, trans_count))
g + geom_boxplot(aes(fill=factor(group)))+ scale_fill_brewer(palette="Set1")+guides(fill=FALSE)+
   labs(x="Female Gut Metagenome",y="log10 of Beta-Glucuronidase Hits")+theme_classic(base_size = 30)+
   geom_point()+scale_x_continuous(breaks=c(1, 2,3),labels=c("Low", "Medium","High"))

#Plotting boxplot of the pecentiles
df$group <- as.numeric(cut(df$trans_count, 2))
g <- ggplot(df, aes(group, trans_count))
g + geom_boxplot(aes(fill=factor(group)))+ scale_fill_brewer(palette="Set1")+guides(fill=FALSE)+
   labs(x="Female Gut Metagenome",y="log10 of Beta-Glucuronidase Hits")+theme_classic(base_size = 30)+
   geom_point()+scale_x_continuous(breaks=c(1,2),labels=c("Low", "High"))

#######################################################
tab=read.csv("MBS.csv",header=T)
NeedMBS5=merge(NeedMBS, tab, all.x = TRUE)
#############################################CODE for getting MBS for SRA# Jose experimenting
entrez_dbs()
entrez_db_summary(db = "sra")
entrez_db_searchable(db = "sra")
run = entrez_search(db = "sra", term = "ERR2009629[ACCN]")

str(run)
run[[1]]

run = entrez_summary(db = "sra", id = run[[1]])
run[1:5]

faa  = entrez_fetch(db = "sra", id = run[[1]], rettype = "xml", parsed =TRUE)
faa

rootNode <- XML:: xmlRoot(faa)
rootNode[1]

data = XML::xmlSApply(rootNode,function(x) XML::xmlSApply(x, XML::xmlValue))
cd.catalog <- data.frame(t(data),row.names=NULL)

XML::xpathSApply(rootNode, "serica", XML::xmlValue)

mlSApply(rootNode,function(x) XML::xmlSApply(x, XML::xmlValue))

## 'getSize' takes a vector containing NCBI SRA accessions and returns and vector 
## with the run size (Mb) 
# Dependencies
library(rentrez)


getSize <- function(ids) {
  
  size_mega <- c()
  
  for(i in ids) {
    
    term =  paste0(i, "[ACCN]")
    run = entrez_search(db = "sra", term = term)
    exp_descrip = entrez_summary(db = "sra", id = run[[1]])
    x = exp_descrip$run
    
    ## extract range defined by "total bases" and "load_done"
    size = substr(x, start = regexpr("total_bases=", x)[[1]][1] + attr(regexpr("total_bases=", x), "match.length"),
                  stop = regexpr(" load_done", x)[[1]][1])
    ### some clean up
    size = gsub('\"', "", size, fixed = TRUE)
    
    
    ## converto to numeric and Mb
    size = as.numeric(size)/1e6
    
    
    size_mega <- c(size_mega, size)
    
  }
  
  size_mega
  
}

## Usage 
ids <- c("DRR071071", "ERR1912953")
getSize(ids)

#Another way to do it#####################################################
library(rentrez)
library(stringr)

getSize <- function(ids) {
  
  sizes <- c()
  
  for(i in ids) {
    
    term =  paste0(i, "[ACCN]")
    run = entrez_search(db = "sra", term = term)
    exp_descrip = entrez_summary(db = "sra", id = run[[1]])
    x = exp_descrip$run
    
    size = str_sub(x, start =str_locate(x, "total_bases=")[2]+2, 
                   end =str_locate(x, "load_done")[1]-3)
    
    size_mega = as.numeric(size)/1e6
    
    sizes <- c(sizes, size_mega)
    
  }
  
  sizes
  
}

```

# Second Script


```r
tab=read.csv("body_sites.csv",header=T)

library(ggplot2)
library(scales)


gg <- ggplot(tab, aes(count)) + 
  geom_point(aes(col=enzyme)) + 
  xlim(c(0, 0.1)) + 
  ylim(c(0, 500000)) + 
  labs(subtitle="Area Vs Population", 
       y="Population", 
       x="Area", 
       title="Scatterplot", 
       caption = "Source: midwest")

plot(gg)

theme_set(theme_classic())

# Plot
ggplot(tab, aes(x=cond, y=count)) + 
  geom_point(col="tomato2", size=3) +   # Draw points
  
  labs(title="Dot Plot", 
       subtitle="Make Vs Avg. Mileage", 
       caption="source: mpg") +  
  coord_flip()  +facet_grid(~enz)

tab=read.csv("body_sites.csv",header=T)
tab=read.csv("graph.csv",header=T)


g <- ggplot(data = subset(tab,enz=="Beta-glucuronidase"), aes(cond, log2))
g + geom_boxplot(aes(fill=factor(cond))) + 
 #theme(axis.text.x = element_text(angle=30, vjust=0.6)) + 
  labs(title="Box plot", x="Body Site",y="Enzymes")#+ facet_wrap( ~ enz)


#individual enzymes
g <- ggplot(data = subset(tab,enz=="Beta-glucuronidase" & cond==c("Healthy Adult","Healthy Infant")), aes(cond, log2))
g + geom_boxplot(aes(fill=factor(cond)))+ scale_fill_brewer(palette="Spectral")+guides(fill=FALSE)+
 # theme(axis.text.x = element_text(angle=30, vjust=0.6)) +
  labs(x="Gut Metagenome",y="Beta-Glucuronidase Hits")+theme_classic(base_size = 30)+ geom_point()

#individual enzymes
g <- ggplot(data = subset(tab,enz=="Glutamate-decarboxylase"), aes(cond, count))
g + geom_boxplot(aes(fill=factor(cond)))+ scale_fill_brewer(palette="Spectral")+guides(fill=FALSE)+
  labs(x="Body Site",y="Glutamate Decarboxylase Hits")+theme_classic(base_size = 30)+ geom_point()




#specifiy iptons
g <- ggplot(data = subset(tab,enz=="Beta-glucuronidase"), aes(cond, log2))
g + geom_boxplot(aes(fill=factor(cond)))+ scale_fill_brewer(palette="Set1")+guides(fill=FALSE)+
  # theme(axis.text.x = element_text(angle=30, vjust=0.6)) +
  labs(x="Body Site Metagenome",y="Beta-Glucuronidase Hits")+theme_classic(base_size = 30)+ geom_point()




res.aov2 <- aov(log2 ~ cond * enz, data = tab)
summary(res.aov2)


ANV=TukeyHSD(res.aov2)
  write.csv(ANV[3],"anova3_bodysites.csv")
str(ANV)

#original
g <- ggplot(tab, aes(cond, count))
g + geom_violin(aes(fill=factor(cond))) + 
  theme(axis.text.x = element_text(angle=65, vjust=0.6)) + 
  labs(title="Box plot", 
       subtitle="City Mileage grouped by Class of vehicle",
       caption="Source: mpg",
       x="Body",
       y="Enzymes") + facet_wrap( ~ enz)
```

# Third Script
```r
## Dependencies
library(rentrez)

## Define function
get_CDDacc <- function(prot_vector) {
   
   # initializes a vector
   acc = vector(mode = "character", length = length(seq(prot_vector)))
   
   # loop over the protein ids list 
   for(i in seq(prot_vector)) {
      
      # NCBI Eutils through rentrez pckg.
      protein <-  entrez_summary(db="protein", id= prot_vector[i])
      p_links = entrez_link(dbfrom='protein', id=protein$uid, db='cdd')
      cdd = p_links$links$protein_cdd_concise_2
      
      # if CDD
      if(! is.null(cdd)) {
         region = entrez_summary(db="cdd", id= cdd)
         acc[i] = paste(region$accession,"_",region$abstract)
      } else {acc[i] = 0}
      
   }
   
   acc  
   
}

## Usage
genome= entrez_summary(db="nuccore", id= "NC_010463.1")
link=entrez_link(dbfrom="nuccore",db = "protein",id= genome$uid)
prot=entrez_fetch(db="protein",id = link$links$nuccore_protein, rettype="fasta")
cat(prot, file="APP_transcripts.fasta")

ppp=link$links$nuccore_protein

xps = c("YP_001949746.1", "YP_001949758", "YP_001949763", "YP_001949766", "YP_001949773", "YP_001949775", "YP_001949776", "YP_001949796", "YP_001949797", "YP_001949798", "YP_004306652", "YP_004306653", "YP_004306654", "YP_004306656", "YP_004306662", "YP_004306669", "YP_004306672", "YP_004306675", "YP_004306679", "YP_004306680", "YP_004306696")
y=as.data.frame(ppp)
y$CDD=get_CDDacc(y$ppp)
library(plyr)
dff <- count(y, 'CDD')
dff <- dff[order(-dff$freq),]

```


# Fourth Script

```r
library(rentrez)

search_year <- function(year, term){
   query <- paste(term, "AND (", year, "[PDAT])")
   entrez_search(db="protein", term=query, retmax=0,)$count
}

year <- 1988:2018
proteins <- sapply(year, search_year, term="Pseudomonas[Organism]", USE.NAMES=FALSE)

#plot(year, log(proteins,10), type='b', main="The Rise of the proteins")

x=as.data.frame(cbind(year,proteins))

library(ggplot2)
ggplot(x, aes(x=year, y=proteins, label=proteins)) + geom_line(linetype = "dashed") +
   geom_point(shape=19,size=10,col="darkred")  +
   geom_text(color="white", size=2)+scale_x_continuous(breaks = seq(1988, 2020, by = 2))+
   theme_classic(base_size = 15)+labs(x="Year",y="Number of Proteins")
```


